src/java/org/apache/cassandra/audit/AuditLogEntry.java

```diff
@@ -0,0 +1,302 @@
+        {
+            builder.append("|operation:").append(operation);
+        }
+        return builder.toString();
+    }
+
+    public InetAddressAndPort getHost()
+    {
+        return host;
+    }
+
+    public InetAddressAndPort getSource()
+    {
+        return source;
+    }
+
+    public String getUser()
+    {
+        return user;
+    }
+
+    public long getTimestamp()
+    {
+        return timestamp;
+    }
+
+    public AuditLogEntryType getType()
+    {
+        return type;
+    }
+
+    public UUID getBatch()
+    {
+        return batch;
+    }
+
+    public String getKeyspace()
+    {
+        return keyspace;
+    }
+
+    public String getScope()
+    {
+        return scope;
+    }
+
+    public String getOperation()
+    {
+        return operation;
+    }
+
+    public QueryOptions getOptions()
+    {
+        return options;
+    }
+
+    public static class Builder
+    {
+        private static final InetAddressAndPort DEFAULT_SOURCE;
+
+        // thanks, java, for making this ugly
+        static
+        {
+            try
+            {
+                DEFAULT_SOURCE = InetAddressAndPort.getByNameOverrideDefaults("0.0.0.0", 0);
+            }
+            catch (UnknownHostException e)
+            {
+
+                throw new RuntimeException("failed to create default source address", e);
+            }
+        }
+
+        private static final String DEFAULT_OPERATION = StringUtils.EMPTY;
+
+        private AuditLogEntryType type;
+        private InetAddressAndPort source;
+        private String user;
+        private long timestamp;
+        private UUID batch;
+        private String keyspace;
+        private String scope;
+        private String operation;
+        private QueryOptions options;
+
+        public Builder(ClientState clientState)
+        {
+            if (clientState != null)
+            {
+                if (clientState.getRemoteAddress() != null)
+                {
+                    InetSocketAddress addr = clientState.getRemoteAddress();
+                    source = InetAddressAndPort.getByAddressOverrideDefaults(addr.getAddress(), addr.getPort());
+                }
+
+                if (clientState.getUser() != null)
+                {
+                    user = clientState.getUser().getName();
+                }
+                keyspace = clientState.getRawKeyspace();
+            }
+            else
+            {
+                source = DEFAULT_SOURCE;
+                user = AuthenticatedUser.SYSTEM_USER.getName();
+            }
+        }
+
+        public Builder(ClientState clientState, long queryStartNanoTime)
+        {
+            this(clientState);
+            /*
+             * FQLTime is time in milliseconds since the epoch since the operation was invoked, this does not involve operation execution time
+             */
+            long fqlTime = System.currentTimeMillis() - TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - queryStartNanoTime);
```

Comment:
@jasobrown What do you think on this calculation? Is alternative to directly rely on `queryStartNanoTime` sgty?

Reply:
@jasobrown ^

Reply:
ughh, two system calls in one line. let me see what I can do (will be part of a patch of changes I'm pushing up in a short while).